### PR TITLE
prober: support adding key/value labels to probes.

### PR DESCRIPTION
**Pull request by danderson**
This lets especially dynamically-registered probes export their results with a bunch more dimensions to slice by in dashboards, so you can e.g. show "only probes for DERP in Singapore", or "only production backup checks".

The labels are still static after probe creation, because prometheus doesn't deal well with per-datapoint cardinality explosion. This API forces you to think a little bit about adding labels that are about the probe, not about a particular probing event.

Two commits. First one is a pure refactor that uses PrometheusVar for exporting probe results, without altering the exported data. The second commit alters the probe export to include custom labels.

### CODE REVIEW

File: prober/prober.go [hunk 1]
```diff
@@ -149,6 +107,11 @@ type Probe struct {
 	doProbe  ProbeFunc
 	interval time.Duration
 	tick     ticker
+
+	mu     sync.Mutex
+	start  time.Time
+	end    time.Time
+	result bool
```
Comment (bradfitz) [inline prober/prober.go:117]: what's result? has a result? has a good result?

**Review by bradfitz — APPROVED** [review]
Other than test failure.

File: prober/prober.go [hunk 2]
```diff
@@ -192,35 +158,120 @@ func (p *Probe) run() {
...
+
+	bs, err := json.Marshal(out)
 	if err != nil {
-		v = 0
+		return fmt.Sprintf(`{"error": %q}`, err)
+	}
+	return string(bs)
+}
+
+func (v varExporter) WritePrometheus(w io.Writer, prefix string) {
```
Comment (maisem) [inline prober/prober.go:234]: way too much going on here to understand. Please add docs?

File: prober/prober.go [hunk 3]
```diff
@@ -192,35 +158,120 @@ func (p *Probe) run() {
...
+	defer p.mu.Unlock()
+	p.end = end
+	p.result = err == nil
+}
+
+type varExporter struct {
+	p *Prober
+}
+
+func (v varExporter) String() string {
```
Comment (maisem) [inline prober/prober.go:195]: same

File: prober/prober.go (hunk 2 above)
Comment (maisem) [reply prober/prober.go:234]: specifically i care about what fields it exports and what the output format looks like so that I can debug it.

Comment (maisem) [reply prober/prober.go:234]: and this should have a test to verify output matches whats expected.

**Comment by danderson** [comment]
Wow, I thought I'd added tests for this. Sigh, mondays. Adding tests and stuff before un-drafting again.

**Comment by danderson** [comment]
Okay, added more comments and tests for the two expvar formats. @maisem PTAL

**Comment by danderson** [comment]
I ended up squashing the changes into the wrong commit of this PR, so I'll do a squash and merge to keep the timeline in main clean. Sorry for the confusion.

**Review by maisem — APPROVED** [review]

**Pull request merged**